### DISCUSSION TITLE
Markdown editor

### DISCUSSION THREAD

**Discussion by danipen** [AvaloniaUI/AvaloniaEdit#224] [Ideas]
Hey @mattj23,

I read you're working on a Markdown editor over AvaloniaEdit. Some questions:

- Do you plan to open source it? In that case, I'd be interested in contributing since we're going to use it.
- Are you using https://github.com/whistyun/Markdown.Avalonia

**Comment by mattj23** [comment]
@danipen,

I'm still in the early stages trying to prove out feasibility for the different parts of what I'm hoping to make, so I have it as a private repository right now, but yes, ultimately my goal is to create something in a reaction to what I see as a proliferation of a bunch of closed source and/or electron-based applications:

* GPL licensed, completely free and open and resistant to commercial exploitation
* heavily extensible with plugins
* visually customizable
* well performing (better than electron)
* cross platform (MacOS, Linux, Windows)
* WYSIWYG-ish, like Zettlr/Typora/Obsidian

The disclaimer, though, is that it's not going to be strictly a traditional editor.  I have a main goal for it and a secondary goal:

1. Principally, I want to build something for people writing long-form prose, like academic/scholarly articles, journalistic essays, books, etc.  This would be an alternative to apps like Scrivner and Ulysses, by incorporating some of those types of features like outlining, visual shuffling of sections, writing targets, interfacing with citation management, and streamlining export to pandoc. 

2. The secondary goal would be to target people doing note-taking and management along the lines of Zettlr, Obsidian, Roam, Bear, and Typora.  This is more straightforward and would mostly consist of managing internal linking, collections of documents, tagging, and having a good plugin system.

If I could accomplish those there would also be a vanilla markdown editor in there, and it could be used as such.

> Are you using https://github.com/whistyun/Markdown.Avalonia

I haven't looked at rendering markdown yet, since it's not strictly necessary for the two goals I have above and also I didn't think it would be a high risk part of the project.  I'm using [Markdig](https://github.com/xoofx/markdig) for parsing and it has a rendering pipeline in it to produce HTML, but I hadn't looked at actually getting that to the screen yet.

> In that case, I'd be interested in contributing since we're going to use it.

If what I described still sounds interesting to you, I can't think of anything that would be better than having the help of someone intimately familiar with AvaloniaEdit, since that's where I've hit the most obstacles in my proof of concept work.  I think every other part of the project I'm pretty confident I can accomplish, but I'm pretty new to Avalonia (I'm a WPF refugee, this has been my only experiment so far with Avalonia: https://github.com/mattj23/BookShuffler, a couple days of work about a year ago).

If you're interested I'll clean up the repo and invite you to it. I'll have some documentation in there about exactly what I'm hoping to accomplish.

**Comment by danipen** [comment]
Thanks you for the info.

Some questions.

Do you plan to render html?https://github.com/whistyun/Markdown.Avalonia allows to tender markdownn using avalonia controls ... not html.

What would you use AvaloniaEdit for?

**Reply by mattj23** [reply]
I had assumed that I'd need to go to HTML before being able to render, but if that step can be skipped I'm open to doing it in whatever the most expedient way is; I don't have any preference myself.  I think in the long run it would be good to be able to extend the rendering system if someone wanted to add a custom markdown type (such as some of the extensions that Pandoc supports), so if Markdown.Avalonia can do that all the better.  But I'm hoping to make the software extensible enough that someone could always replace these parts of it with something else if they don't like it.

What I'm trying to do with AvaloniaEdit is make the text editor for the markdown source itself with a WYSIWYG-like interface, the sort of way that Zettlr or Typora do, in which they'll optionally remove certain characters (like the `*` emphasis, for example) when the cursor isn't in that node of the abstract syntax tree, replacing it with just the bold or italicized text. For a lot of people it provides a reasonable compromise to having a live preview running next to the markdown source taking up screen space, though I'd also like to have the live preview as an option at some point.

So in that sense I'm trying to use AvaloniaEdit for the main editing with some additional niceties like replacing citekeys, links, and footnotes with folding-like representations that move that information elsewhere in the interface without cluttering up the markdown source.

**Comment by PeteMichaud** [comment]
@mattj23 I found this thread while doing research for essentially the same goal you have, so maybe we can team up. I'm also new to avalonia, however, so temper expectations accordingly :)

**Reply by mattj23** [reply]
@PeteMichaud, I'd welcome any help, and conveniently we're located fairly close to each other (I spend a lot of time in Philly).  Let me get a little bit further on a skeleton of the project and if you're still interested I'll invite you to it?

**Reply by danipen** [reply]
Any updates on this project?

**Comment by fentazyz** [comment]
请问有遇到过在使用AvaloniaEdit的Document.Insert方法插入文本之后就无法输入中文的问题吗
